File: Week_04/num_islands.rs
impl Solution {
    pub fn num_islands(grid: Vec<Vec<char>>) -> i32 {
        let rows = grid.len();
        if rows == 0 {return 0;}
        let cols = grid[0].len();
        let mut editable_grid = grid;
        let mut ans: i32 = 0;

        for row in 0..rows {
            for col in 0..cols {
                if editable_grid[row][col] == '1' {
                    ans += 1;
                    Self::dfs(row, col, &mut editable_grid);
                }
            }
        }
        ans
    }

    fn dfs(row: usize, col: usize, grid: &mut Vec<Vec<char>>) {
        if row < 0 || col < 0 || row >= grid.len() || col >= grid[0].len() || grid[row][col] == '0' {return;}

        grid[row][col] = '0';
        // 上
        Self::dfs(row - 1, col, grid);
        // 右
        Self::dfs(row, col + 1, grid);
        // 下
        Self::dfs(row + 1, col, grid);
        // 左
        Self::dfs(row, col - 1, grid);
    }
}

File: Week_07/numIslands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        rows, ans = len(grid), 0
        if rows == 0: return 0
        cols = len(grid[0])
        for row in range(rows):
            for col in range(cols):
                if grid[row][col] == '0': continue
                ans += 1
                self.dfs(row, col, grid)
        return ans

    def dfs(self, row, col, grid):
        if -1 < row < len(grid) and -1 < col < len(grid[0]) and grid[row][col] == '1':
            grid[row][col] = '0'
            for x, y in ((1, 0), (0, 1), (-1, 0), (0, -1)): self.dfs(row + x, col + y, grid)